tests/test_cleanup.py

```diff
@@ -487,3 +507,808 @@ def test_file_vdi_delete(self, mock_lock):
+
+    @mock.patch('cleanup.VDI', autospec=True)
+    @mock.patch('cleanup.SR._liveLeafCoalesce', autospec=True)
+    @mock.patch('cleanup.SR._snapshotCoalesce', autospec=True)
+    def test_coalesceLeaf_coalesce_failed(self,
+                                          mock_srSnapshotCoalesce,
+                                          mock_srLeafCoalesce,
+                                          mock_vdi):
+
+        mock_vdi.canLiveCoalesce.return_value = False
+        mock_srSnapshotCoalesce.return_value = False
+        mock_srLeafCoalesce.return_value = False
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        res = sr._coalesceLeaf(vdi)
+        self.assertFalse(res)
+
+    @mock.patch('cleanup.VDI.canLiveCoalesce',
+                autospec=True, return_value=False)
+    @mock.patch('cleanup.VDI.getSizeVHD',
+                autospec=True, return_value=1024)
+    @mock.patch('cleanup.Util.log')
+    @mock.patch('cleanup.SR._snapshotCoalesce',
+                autospec=True, return_value=True)
+    def test_coalesceLeaf_size_the_same(self,
+                                        mock_srSnapshotCoalesce,
+                                        mock_log,
+                                        mock_vdisize,
+                                        mockliveCoalesce):
+
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        with self.assertRaises(util.SMException) as exc:
+            res = sr._coalesceLeaf(vdi)
+
+        self.assertEqual("VDI {uuid} could not be"
+                         " coalesced".format(uuid=vdi_uuid),
+                         exc.exception.message)
+
+    @mock.patch('cleanup.VDI.canLiveCoalesce', autospec=True,
+                return_value=False)
+    @mock.patch('cleanup.VDI.getSizeVHD', autospec=True)
+    @mock.patch('cleanup.SR._snapshotCoalesce', autospec=True,
+                return_value=True)
+    @mock.patch('util.SMException.__init__', autospec=True, return_value=None)
+    @mock.patch('cleanup.Util.log')
+    def test_coalesceLeaf_size_bigger(self, mock_log, mock_exceptionInit,
+                                      mock_snapshotCoalesce, mock_vhdSize,
+                                      mock_vdiLiveCoalesce):
+
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        mock_vhdSize.side_effect = iter([1024, 4096, 4096, 8000, 8000, 16000])
+
+        sr._snapshotCoalesce = mock.MagicMock(autospec=True)
+        sr._snapshotCoalesce.return_value = True
+
+        with self.assertRaises(util.SMException) as exc:
+            res = sr._coalesceLeaf(vdi)
+
+        mock_exceptionInit.assert_called_with("VDI {uuid} could"
```

Comment:
This should be the same as the method above?

Reply:
dang

Reply:
fixed